sycl/doc/extensions/SubGroupBitmaskGroupFunctions/sub_group_bitmask_group_algorithms.asciidoc

```diff
@@ -0,0 +1,253 @@
+divergence is why work-items following the same control flow may not be
+active at the same time. Care should be taken when modifying the value of the 
+returned `sub_group_mask` since this will change the behavior of masked group functions.
+
+`ext_oneapi_rank_in_mask` returns an integer id representing the calling work-item's
+position within sub-group subset represented by the provided `sub_group_mask`. This is
+undefined behavior if bit `n` of the `sub_group_mask` is not set, where `n` is the
+integer id of the work-item in the sub-group. As an example, if the bitmask was 
+`0b00000000 00000000 11110000 11110000` and `get_sub_group().get_local_linear_id() == 13`,
+ext_oneapi_rank_in_mask would return 5.
+
+
+```c++
+namespace sycl {
+template <int dimensions = 1>
+class nd_item {
+  public:
+...
+  // create a bitmask of the sub-group subset this work-item would be a part of
+  // if the sub-group was split into contiguous groups of size `partition_size`
+  sub_group_mask ext_oneapi_partition_sub_group(size_t partition_size) const;
+
+  // get the bitmask representing the work-items of the sub-group that are
+  // actively executing.
+  sub_group_mask ext_oneapi_active_sub_group_items() const;
+
+  //Returns an integer id representing the calling work-item's position within the provided sub_group_mask.
+  size_t ext_oneapi_rank_in_mask(sub_group_mask mask) const;
+...
+};
+}
+```
+
+possible implementation of `ext_oneapi_rank_in_mask`:
+```c++
+size_t ext_oneapi_rank_in_mask(sub_group_mask mask){
+  // taking 1 from a power of two will give all 1s below the bit of our work-item
+  // anding this with the mask will leave only 1s where there are work-items with lower ids
+  // the popcount of this is the linear id in that sub_group_mask
+  return popcount(mask & ((1 << get_sub_group().get_local_linear_id()) -1));
+}
+```
+
+## Changes to `joint_reduce` and `reduce_over_group`
+Addition of bitmasked version of methods for sub-groups.
+
+```c++
+namespace sycl::ext::oneapi {
+template <typename Ptr, class BinaryOperation>
+joint_reduce(sycl::sub_group, Ptr first, Ptr last, BinaryOperation binary_op, sycl::sub_group_mask);
+
+template <typename Ptr, typename T, class BinaryOperation>
+joint_reduce(sycl::sub_group, Ptr first, Ptr last, T init, BinaryOperation binary_op, sycl::sub_group_mask);
+
+template <typename T, class BinaryOperation>
+reduce_over_group(sycl::sub_group, T x, BinaryOperation binary_op, sycl::sub_group_mask);
+
+template <typename V, typename T, class BinaryOperation>
+reduce_over_group(sycl::sub_group, V x, T init, BinaryOperation binary_op, sycl::sub_group_mask);
+} // namespace sycl::ext::oneapi
+```
+
+## Group_barrier
+New overloads will be required for `sub_group` that takes a bitmask.
+
+```c++
+namespace sycl::ext::oneapi {
+group_barrier(sycl::sub_group, sycl::sub_group_mask mask, memory_scope fence_scope = sub_group::fence_scope);
+} // namespace sycl::ext::oneapi
+```
+
+## New `async_group_copy` function.
```

Comment:
Aside: This would uphold the precedent for SYCL functions having an `async` prefix for asynchronous algorithms.  Do you foresee a need for any more, or is `group_copy` the only one you think we need?

I'm a little worried this could lead to families of algorithms like `joint_reduce`, `reduce_over_group`, `async_joint_reduce` and `async_reduce_over_group`, which is a lot of names.

Reply:
I don't find it too unlikely that future architectures could have some kind of asynchronous reduce that ships the data off to another module on the device and if that is the case then I would like the choice. I guess asynchronous versions would have different properties like throughput, latency, occupancy, affect on the cache, and if they return an event in code.

Saying that, I don't think it is just about to happen and `group_copy` is the only one I know about for now.
My only concern about naming it `group_copy` is that it will likely still return an event, which may seem odd, but I'm not that bothered.